File: src/Demodulator.hs
module Demodulator where

import Invaluator (Data(..))
import Debug.Trace (traceShowId)

demodulate :: String -> Data
demodulate input =
  let (value, _) = demodulateInternal input in
  value

demodulateInternal :: String -> (Data, String)
demodulateInternal input =
  case input of
    '0':'0':tail -> (DNil, tail)
    '1':'1':tail ->
      let (first, tail') = demodulateInternal tail
          (second, tail'') = demodulateInternal tail' in
      (DCons first second, tail'')
    '0':'1':tail -> demodulateNumber tail 1
    '1':'0':tail -> demodulateNumber tail (-1)
    _ -> error $ "WTF encoding: " ++ input

demodulateNumber :: String -> Integer -> (Data, String)
demodulateNumber input sign =
  let (number, tail) = demodulateNumberInternal input in
  (DNum $ sign * number, tail)

demodulateNumberInternal :: String -> (Integer, String)
demodulateNumberInternal input =
  case input of
    '0':tail -> (0, tail)
    '1':_ ->
      let (bound, tail') = traceShowId $ demodulateNumberBound input in
      demodulateBoundedNumber (bound `div` 2) tail'
  where
    demodulateNumberBound input =
      case input of
        '1':tail ->
          let (b, tail') = demodulateNumberBound tail in
          (16*b, tail')
        '0':tail -> (1, tail)

demodulateBoundedNumber :: Integer -> String -> (Integer, String)
demodulateBoundedNumber bound input =
  case bound of
    0 -> (0, input)
    _ -> case input of
          '1':tail ->
            let (result, tail') = demodulateBoundedNumber (bound `div` 2) tail in
            (bound + result, tail')
          '0':tail -> demodulateBoundedNumber (bound `div` 2) tail

File: csmodulator/Modulator/Modulator.Core/Demodulator.cs
using System;

namespace Modulator.Core
{
    public class Demodulator
    {
        public static IEntry Demodulate(string input)
        {
            var ptr = 0;
            return DemodulateInternal(input, ref ptr);
        }

        private static IEntry DemodulateInternal(string input, ref int ptr)
        {
            var prefix = input.Substring(ptr, 2);
            ptr += 2;
            switch (prefix)
            {
                case "00":
                    return new Null();
                case "11":
                    var first = DemodulateInternal(input, ref ptr);
                    var second = DemodulateInternal(input, ref ptr);
                    return new Pair(first, second);
                case "01":
                    return new Number(DemodulateNumber(input, ref ptr));
                case "10":
                    return new Number(-DemodulateNumber(input, ref ptr));
                default:
                    throw new Exception("wtf");
            }
        }

        private static int DemodulateNumber(string input, ref int ptr)
        {
            if (input[ptr] == '0')
            {
                ptr++;
                return 0;
            }

            var bound = 1;
            while (input[ptr] == '1')
            {
                bound *= 16;
                ptr++;
            }

            bound /= 2;
            var result = 0;
            ptr++;
            while (bound > 0)
            {
                if (input[ptr] == '1')
                    result += bound;
                ptr++;
                bound /= 2;
            }

            return result;
        }
    }
}
